content/2023-03-03-sgx-block-building.mdx

```diff
@@ -0,0 +1,104 @@
+
+- Fully working [Flashbots block builder](https://github.com/flashbots/builder) (branch [`capella-test`](https://github.com/flashbots/builder/tree/capella-test))
+- [Remote attestation](https://en.wikipedia.org/wiki/Trusted_Computing#Remote_attestation) support, allowing anyone to verify the SGX proof:
+    - A remote-attestation TLS (RA-TLS) certificate is created on startup, which embeds the SGX attestation report.
+    - See [here](https://github.com/CCC-Attestation/meetings/blob/main/materials/ShanweiCen_Interoperable_ATLS.pdf) and [here](https://github.com/gramineproject/gramine/pull/1039) for more information about RA-TLS, as well as the [patch for adding it to the builder](https://github.com/flashbots/geth-sgx-gramine/pull/9/files#diff-c0fdcc479c0e7e75defff1b32dbc6d01e39a7d248d4dbcb6709220b85e68c88b) codebase.
+    - RA-TLS is used by several SGX projects including Gramine, Open Enclave and Inclavare Containers.
+- Support for [EDMM](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-edmm) (enclave dynamic memory management) via an update to [Gramine 1.4](https://github.com/gramineproject/gramine/releases/tag/v1.4). EDMM obviates the need to reserve and write all memory on startup, which has the potential to significantly speed up the startup process.
+- [Guide](https://github.com/flashbots/geth-sgx-gramine#attest-enclave) for attesting the certificate of an SGX builder
+- Reproducible builder Docker image, arriving at a reproducible [MRENCLAVE](https://www.intel.com/content/www/us/en/developer/articles/technical/introduction-to-intel-sgx-sealing.html) value when rebuilding the builder Docker image.
+
+
+## Challenges
+
+### Geth state size
+
+At the time of writing, the geth database requires about 1TB of storage.
+
+- The database needs to be encrypted and copied into the enclave on every startup. Given the full mainnet geth database with about one terabyte, the startup process takes about 4.5 hours.
+- This also means it takes a long time to reproduce bugs.
+- Gramine is crashing at 1TB of state in the enclave (we’ve opened a [bug report](https://github.com/gramineproject/gramine/issues/1207))
+- A promising direction is this [geth PBSS patch](https://github.com/ethereum/go-ethereum/pull/26274) (auto-pruning storage) which might be able to reduce the state size to about 150GB (see also [this post](https://nodereal.io/blog/en/geth-path-based-storage-model-and-newly-inline-state-prune/) for more details)
+
+### Merging performance and additional latencies
+- The SGX builder currently performs at around ~150 mgasps, which is about half the performance of a non-SGX builder. This means that the SGX builder can only do half the simulations.
+- A likely culprit for this degradation is the syscalls interface. Possible improvements could be made by building geth with [gccgo](https://go.dev/doc/install/gccgo), for improved performance / raw syscalls vs syscalls via libc.
```

Comment:
A likely culprit for this degradation is the way Golang handles syscalls. Instead of using the `libc` syscall interface Golang calls directly into the kernel. This produces additional execution overhead when running inside an SGX enclave. Possible improvments...

Reply:
thx!